docs/api-docs.rst

```diff
@@ -1,6 +1,217 @@
+important field of the task is the ``mode``, which describes what type of task
+this is. The structure for the task dictionary is described below. Uppercase names
+within angle braces represent symbols defined in the specification. Optional parts
+of the specification are surrounded by parenthesis to avoid ambiguity with the
+square braces, which represent lists in python or Arrays in JSON.
+
+.. code-block :: none
+
+    <TASK> ::= <PYTHON_TASK> | <R_TASK> | <DOCKER_TASK> | <WORKFLOW_TASK>
+
+    <PYTHON_TASK> ::= {
+        "mode": "python",
+        "script": <python code to run as a string>
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <R_TASK> ::= {
+        "mode": "r",
+        "script": <r code to run (as a string)>
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <DOCKER_TASK> ::= {
+        "mode": "docker",
+        "docker_image": <docker image name to run>
+        (, "container_args": [<container arguments>])
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <WORKFLOW_TASK> ::= {
+        "mode": "workflow",
+        "steps": [<WORKFLOW_STEP> (, <WORKFLOW_STEP>, ...)],
+        "connections": [<WORKFLOW_CONNECTION> (, <WORKFLOW_CONNECTION>, ...)]
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <WORKFLOW_STEP> ::= {
+        "name": <step name>,
+        "task": <TASK>
+    }
+
+    <WORKFLOW_CONNECTION> ::= {
+        ("name": <name of top-level input to bind to>)
+        (, "input": <input id to bind to for a step>)
+        (, "input_step": <input step name to connect>)
+        (, "output_step": <output step name to connect>)
+    }
+
+The workflow mode simply allows for a directed acyclic graph of tasks to be
+specified to :py:func:`romanesco.run`.
+
+.. code-block:: none
+
+    <TASK_INPUT> ::= {
+        "id": <string, the variable name>,
+        "type": <data type>,
+        "format": <data format>
+        (, "target": <INPUT_TARGET_TYPE>)   ; default is "memory"
+        (, "filename": <name of file if target="filepath">)
+    }
+
+    <INPUT_TARGET_TYPE> ::= "memory" | "filepath"
+
+    <TASK_OUTPUT> ::= {
+        "id": <string, the variable name>,
+        "type": <data type>,
+        "format": <data format>
+        (, "target": <INPUT_TARGET_TYPE>)   ; default is "memory"
+    }
+
+
+The input specification
+***********************
+
+The ``inputs`` argument to :py:func:`romanesco.run` specifies the inputs to the
+task described by the ``task`` argument. Specifically, it tells what data should
+be placed into the task input ports.
+
+.. code-block:: none
+
+    <INPUTS> ::= {
+        <id> : <INPUT_BINDING>
```

Comment:
Should probably be `[<id>: <INPUT_BINDING> (, <id> : <INPUT_BINDING>, ...)]` to show there can be multiple inputs.